File: cfid_go_if/ifreceipt_test.go
package ifreceipt

import "testing"

// colorfulPixels returns the 4x4 canonical test image from SPEC-IF-1.0.md A.1.
func colorfulPixels() [][3]int {
   colors := [][3]int{{220,50,50},{50,180,80},{50,90,220},{240,200,40}}
   pixels := make([][3]int, 16)
   for i := range pixels {
   pixels[i] = colors[(i%4+i/4)%4]
   }
   return pixels
}

func TestCanonicalPixelDigestVector(t *testing.T) {
   got := PixelDigestCanonical(colorfulPixels())
   want := "sha256:2a40854db5a950bbdc8d921620a2ee8074fb8b950102150b6be7b10990e1ddb6"
   if got != want {
   t.Errorf("pixel_digest = %s, want %s", got, want)
   }
}

func TestHash12(t *testing.T) {
   got := Hash12("sha256:2a40854db5a950bbdc8d921620a2ee8074fb8b950102150b6be7b10990e1ddb6")
   if got != "2A40854DB5A9" {
   t.Errorf("Hash12 = %s, want 2A40854DB5A9", got)
   }
}

func TestIFIDFormat(t *testing.T) {
   pd := "sha256:2a40854db5a950bbdc8d921620a2ee8074fb8b950102150b6be7b10990e1ddb6"
   rd := "sha256:" + "0000000000000000000000000000000000000000000000000000000000000000"
   id := "IF-" + Hash12(pd) + "-" + Hash12(rd)
   if len(id) != 28 {
   t.Errorf("IF-ID length = %d, want 28", len(id))
   }
   if id[:3] != "IF-" {
   t.Errorf("IF-ID does not start with IF-")
   }
}

func TestMakeReceiptIsValid(t *testing.T) {
   fields := DefaultFields(colorfulPixels(),
   "sha256:"+"0000000000000000000000000000000000000000000000000000000000000000",
   "png", 4, 4)
   r := MakeReceipt(fields)
   result := VerifyReceipt(r)
   if !result.Valid {
   t.Errorf("receipt invalid: %s", result.Reason)
   }
}

func TestSamePixelsSameDigest(t *testing.T) {
   if PixelDigestCanonical(colorfulPixels()) != PixelDigestCanonical(colorfulPixels()) {
   t.Error("same pixels produced different digest")
   }
}

func TestDifferentPixelsDifferentDigest(t *testing.T) {
   red := [][3]int{{255,0,0},{255,0,0},{255,0,0},{255,0,0}}
   blue := [][3]int{{0,0,255},{0,0,255},{0,0,255},{0,0,255}}
   if PixelDigestCanonical(red) == PixelDigestCanonical(blue) {
   t.Error("different pixels produced same digest")
   }
}

func TestTamperedReceiptDetected(t *testing.T) {
   fields := DefaultFields(colorfulPixels(),
   "sha256:"+"0000000000000000000000000000000000000000000000000000000000000000",
   "png", 4, 4)
   r := MakeReceipt(fields)
   r.IFID = "IF-000000000000-000000000000"
   result := VerifyReceipt(r)
   if result.Valid {
   t.Error("tampered receipt should be invalid")
   }
}

func TestParentIFIDNilForOriginal(t *testing.T) {
   fields := DefaultFields([][3]int{{255,0,0}},
   "sha256:"+"0000000000000000000000000000000000000000000000000000000000000000",
   "png", 1, 1)
   r := MakeReceipt(fields)
   if r.ParentIFID != nil {
   t.Error("ParentIFID should be nil for original")
   }
}

func TestIFVersion(t *testing.T) {
   fields := DefaultFields([][3]int{{255,0,0}},
   "sha256:"+"0000000000000000000000000000000000000000000000000000000000000000",
   "png", 1, 1)
   r := MakeReceipt(fields)
   if r.IFVersion != "IF-CAPTURE-1.0.0" {
   t.Errorf("IFVersion = %s, want IF-CAPTURE-1.0.0", r.IFVersion)
   }
}

File: cfid_kotlin_if/src/test/kotlin/com/imagecapture/ifreceipt/IFReceiptTest.kt
package com.imagecapture.ifreceipt

import kotlin.system.exitProcess

private var passed = 0
private var failed = 0

private fun check(name: String, ok: Boolean) {
    if (ok) { println("  ok  $name"); passed++ }
    else    { println("  FAIL $name"); failed++ }
}

private fun colorfulPixels(): List<Triple<Int, Int, Int>> {
    val colors = listOf(
        Triple(220,50,50), Triple(50,180,80),
        Triple(50,90,220), Triple(240,200,40)
    )
    return (0 until 16).map { i -> colors[(i % 4 + i / 4) % 4] }
}

fun main() {
    check("canonical pixel_digest vector: colorful 4x4 PNG",
        pixelDigestCanonical(colorfulPixels()) ==
        "sha256:2a40854db5a950bbdc8d921620a2ee8074fb8b950102150b6be7b10990e1ddb6"
    )

    check("hash12 extracts 12 uppercase hex chars",
        hash12("sha256:2a40854db5a950bbdc8d921620a2ee8074fb8b950102150b6be7b10990e1ddb6") == "2A40854DB5A9"
    )

    val pd = "sha256:2a40854db5a950bbdc8d921620a2ee8074fb8b950102150b6be7b10990e1ddb6"
    val rd = "sha256:" + "0".repeat(64)
    val testId = "IF-" + hash12(pd) + "-" + hash12(rd)
    check("IF-ID format: 28 chars, starts with IF-", testId.length == 28 && testId.startsWith("IF-"))

    val fields = IFReceiptFields(
        pixels = colorfulPixels(), byteDigest = "sha256:" + "0".repeat(64),
        format = "png", width = 4, height = 4
    )
    val receipt = makeReceipt(fields)
    check("makeReceipt produces a valid receipt", verifyReceipt(receipt).valid)

    check("same pixels -> same pixel_digest",
        pixelDigestCanonical(colorfulPixels()) == pixelDigestCanonical(colorfulPixels())
    )

    val red = List(4) { Triple(255, 0, 0) }
    val blue = List(4) { Triple(0, 0, 255) }
    check("different pixels -> different pixel_digest", pixelDigestCanonical(red) != pixelDigestCanonical(blue))

    val tampered = receipt.copy(if_id = "IF-000000000000-000000000000")
    check("tampered if_id is detected", !verifyReceipt(tampered).valid)

    check("parent_if_id is null for original", receipt.parent_if_id == null)

    check("if_version is IF-CAPTURE-1.0.0", receipt.if_version == "IF-CAPTURE-1.0.0")

    println("\n$passed passed, $failed failed")
    if (failed > 0) exitProcess(1)
}
